### PR TITLE
boot,partition: improve tests/docs around SetNextBoot()

### PR DESCRIPTION
**Pull request by mvo5**
This implements the suggested improvements from PR  #5145.

### CODE REVIEW

File: boot/kernel_os.go [hunk 1]
```diff
@@ -134,10 +134,10 @@ func SetNextBoot(s *snap.Info) error {
 		return err
 	}
 	if m[goodBoot] == blobName {
-		// If we were in "try" mode before and now switch to
-		// the good core/kernel again, make sure to clean the
-		// snap_mode here. This also mitigates
-		// https://forum.snapcraft.io/t//5253
+		// If we were in anyting but default ("") mode before
```
Comment (zyga) [inline boot/kernel_os.go]: typo: `s/anyting/anything/`

File: partition/bootloader.go [hunk 2]
```diff
@@ -119,7 +119,23 @@ func ForceBootloader(booloader Bootloader) {
 
 // MarkBootSuccessful marks the current boot as successful. This means
 // that snappy will consider this combination of kernel/os a valid
-// target for rollback
+// target for rollback.
+//
+// The states that a boot goes through are the following:
+// - By default snap_mode is "" in which case the bootloader loads
+//   the snap_core and snap_kernel.
```
Comment (zyga) [inline partition/bootloader.go]: > By default `snap_mode` is `""` in which case the bootloader loads two squashfs'es denoted by variables `snap_core` and `snap_kernel`.

File: partition/bootloader.go [hunk 3]
```diff
@@ -119,7 +119,23 @@ func ForceBootloader(booloader Bootloader) {
...
+// - On a refresh of core/kernel snapd will set snap_mode=try and
+//   will also set snap_try_{core,kernel} to the core/kernel that
+//   will be tried next.
+// - On reboot the bootloader will inspect the snap_mode and if the
+//   mode is set to "try" it will set "snap_mode=trying" and then
+//   try to boot the snap_try_{core,kernel}".
+// - On a (successful) boot snapd will set snap_mode="" (default)
+//   and also snap_{core,kernel} is set to the snap_try_{core,kernel}
```
Comment (zyga) [inline partition/bootloader.go]: On a successful boot snapd resets `snap_mode` to `""` and copies `snap_try_{core,kernel}` to `snap_{core,kernel}`.

Question, do we reset `snap_try_{core,kernel}` at this stage or are they left as-is?

**Comment by mvo5** [comment]
Fwiw, I also added this to the forum: https://forum.snapcraft.io/t/boot-modes-in-ubuntu-core/5464

File: boot/kernel_os.go [hunk 4]
```diff
@@ -134,10 +134,10 @@ func SetNextBoot(s *snap.Info) error {
...
 	}
 	if m[goodBoot] == blobName {
-		// If we were in "try" mode before and now switch to
-		// the good core/kernel again, make sure to clean the
-		// snap_mode here. This also mitigates
-		// https://forum.snapcraft.io/t//5253
+		// If we were in anything but default ("") mode before
+		// and now switch to the good core/kernel again, make
+		// sure to clean the snap_mode here. This also
+		// mitigates https://forum.snapcraft.io/t//5253
```
Comment (bboozzoo) [inline boot/kernel_os.go]: `https://forum.snapcraft.io/t//5253`, notice double `/`, it works but looks a bit off.

Comment (mvo5) [reply boot/kernel_os.go]: Good catch, thank you!

**Pull request merged**